scripts/tools/metricProgrammable.yml

```diff
@@ -0,0 +1,490 @@
+      name: "value"
+      desc: "[in] parameter value"
+--- #--------------------------------------------------------------------------
+type: function
+desc: "Query and get the available metric programmable handles."
+version: "1.8"
+class: $tMetricProgrammable
+name: GetExp
+decl: static
+details:
+    - "Query the available programmable handles using *pCount = 0."
+    - "Returns all programmable metric handles available in the device."
+    - "The application may call this function from simultaneous threads."
+params:
+    - type: "$t_device_handle_t"
+      name: hDevice
+      desc: "[in] handle of the device"
+    - type: uint32_t*
+      name: pCount
+      desc: |
+            [in,out] pointer to the number of metric programmable handles.
+            if count is zero, then the driver shall update the value with the total number of metric programmable handles available.
+            if count is greater than the number of metric programmable handles available, then the driver shall update the value with the correct number of metric programmable handles available.
+    - type: "$t_metric_programmable_exp_handle_t*"
+      name: phMetricProgrammables
+      desc: |
+            [in,out][optional][range(0, *pCount)] array of handle of metric programmables.
+            if count is less than the number of metric programmables available, then driver shall only retrieve that number of metric programmables.
+--- #--------------------------------------------------------------------------
+type: function
+desc: "Get the properties of the metric programmable."
+version: "1.8"
+class: $tMetricProgrammable
+name: GetPropertiesExp
+decl: static
+details:
+    - "Returns the properties of the metric programmable."
+params:
+    - type: $t_metric_programmable_exp_handle_t
+      name: hMetricProgrammable
+      desc: "[in] handle of the metric programmable"
+    - type: "$t_metric_programmable_exp_properties_t*"
+      name: pProperties
+      desc: "[out] properties of the metric programmable"
+--- #--------------------------------------------------------------------------
+type: function
+desc: "Get the information about the parameters of the metric programmable."
+version: "1.8"
+class: $tMetricProgrammable
+name: GetParamInfoExp
+decl: static
+details:
+    - "Returns information about the parameters of the metric programmable handle."
+params:
+    - type: $t_metric_programmable_exp_handle_t
+      name: hMetricProgrammable
+      desc: "[in] handle of the metric programmable"
+    - type: "uint32_t*"
+      name: pParameterCount
+      desc: |
+            [in,out] count of the parameters to retrieve parameter info.
+            if value pParameterCount is greater than count of parameters available, then pParameterCount will be updated with count of parameters available.
+    - type: "$t_metric_programmable_param_info_exp_t*"
+      name: pParameterInfo
+      desc: |
+            [out][range(1, *pParameterCount)] array of parameter info.
```

Comment:
- I saw that **all** APIs that have arrays that are **always** outputs are labeled as [in,out]. @wdamon-intel  would you please clarify why?
- also described as [optional] because you may submit nullptr when querying for the count. 

Above comments applies to all output arrays in the PR.

Reply:
Like mentioned in the previous comment, this is not an optional parameter.
Application would pass in the pointer allocated according to pParameterCount

Reply:
all output arrays that can accept nullptr are described in the spec as optional.

Reply:
@matcabral regarding the `[in,out][optional][range(1, ...)]` on array parameters in APIs such as this, here's a general set of guidelines (note these are not strictly official definitions).

- `in`: the parameter represents an immutable object supplied and managed by the user
- `out`: the parameter represents a mutable object that may be modified and/or managed by the implementation
- `in,out`: the parameter represents an immutable reference to a mutable object; the contents of the referenced object may change as a side effect of the API
- `optional`: the parameter may be `NULL`, the validation layer will not issue a warning or error and the implementation is expected to anticipate this condition.
- `range(x, y)`: the parameter represents a container object with a minimum of _x_ elements and a maximum of _y_ elements, indexable range is [_(x - 1), (y - 1)_]

Reply:
@wdamon-intel Thanks for clarifying. In general I propose to have a common Reference section where all this details are described. This is a source of confusion. 

As a side topic, the fact that the distinction of in and out is made based on the object and not the information/value contained in the object seems contradictory. For instance, IMHO I would expect that "in" is for pieces of data the user feeds to the API for a given purpose and "out" is a piece of data expected as a result of an action of the API. I would expect that params that are "in and out" are for cases in which the user value input is in fact used (as a guidance) and the API updates it to the actual value used (like https://spec.oneapi.io/level-zero/latest/tools/api.html#_CPPv4N26zet_metric_streamer_desc_t14samplingPeriodE).  Labeling [in,out] parameters where the user only inputs the "container" (with no meaningful data for the API) where the API only provides valid output  content should be labeled as [out] (as in this PR). Especially because conflicts with params where there is effectively data going in and out.

Reply:
@joshuaranjan based on the above rule, this (and all other arrays) would need to be in,out

Reply:
@matcabral 
Yes .. Let me make that change

@wdamon-intel Thanks for the explanation.
If we could document this somewhere (maybe a contributing.md), it would help to be consistent across APIs

Reply:
Updated all arrays with [in, out].